src/libmongoc/tests/test-mongoc-client-side-encryption.c

```diff
@@ -4902,6 +4903,101 @@ test_kms_callback (void *unused)
    mongoc_client_destroy (cl);
 }
 
+static void
+_test_auto_aws (bool should_succeed)
+{
+   // Datakey options for AWS
+   mongoc_client_encryption_datakey_opts_t *dk_opts =
+      mongoc_client_encryption_datakey_opts_new ();
+   mongoc_client_encryption_datakey_opts_set_masterkey (
+      dk_opts,
+      tmp_bson ("{ 'region': 'us-east-1', 'key': "
+                "'arn:aws:kms:us-east-1:579766882180:key/"
+                "89fcc2c4-08b0-4bd9-9f25-e30687b580d0' }"));
+
+   // Create a client encryption object
+   mongoc_client_encryption_opts_t *opts = mongoc_client_encryption_opts_new ();
+   mongoc_client_t *cl = test_framework_new_default_client ();
+   mongoc_client_encryption_opts_set_keyvault_client (opts, cl);
+
+   // Given it an on-demand 'local' provider
+   bson_t *empty_aws = tmp_bson ("{'aws': {}}");
+   mongoc_client_encryption_opts_set_kms_providers (opts, empty_aws);
+   mongoc_client_encryption_opts_set_keyvault_namespace (
+      opts, "testing", "testing");
+
+   {
+      // Attempting to create a key from 'aws' will requires credentials in the
+      // environment immediately Create a client encryption object for it.
+      bson_error_t error;
+      mongoc_client_encryption_t *enc =
+         mongoc_client_encryption_new (opts, &error);
+      ASSERT_OR_PRINT (enc, error);
+
+      bson_value_t keyid;
+      mongoc_client_encryption_create_datakey (
+         enc, "aws", dk_opts, &keyid, &error);
+      mongoc_client_encryption_destroy (enc);
+
+      if (should_succeed) {
+         bson_value_destroy (&keyid);
+         ASSERT_OR_PRINT (error.code == 0, error);
+      } else {
+         // We should encounter an error while attempting to connect to the EC2
+         // metadata server.
+         ASSERT_ERROR_CONTAINS (error,
+                                MONGOC_ERROR_CLIENT,
+                                MONGOC_ERROR_CLIENT_AUTHENTICATE,
+                                "failed to contact EC2 link local server");
+      }
+   }
+
+   mongoc_client_encryption_datakey_opts_destroy (dk_opts);
+   mongoc_client_encryption_opts_destroy (opts);
+   mongoc_client_destroy (cl);
+}
+
+static void
+test_auto_aws_fail (void *unused)
+{
+   _test_auto_aws (false);
+}
+
+static void
+test_auto_aws_succeed (void *unused)
+{
+   _test_auto_aws (true);
+}
+
+static int
+_have_aws_creds_env (void *unused)
```

Comment:
I struggle with how these are supposed to be named, since they don't "skip" the test, but rather check some condition, and the test runner decides that a "false" value means to skip the test. Naming is hard.